desc/experimental/_terpsichore.py

```diff
@@ -0,0 +1,872 @@
+        gmns[0, :] = 0
+        gmns[1:, :] = -s
+
+    # |B|
+    bmnc = file.createVariable("bmnc", np.float64, ("radius", "mn_mode_nyq"))
+    bmnc.long_name = "cos(m*t-n*p) component of |B|, on half mesh"
+    bmnc.units = "T"
+    if not eq.sym:
+        bmns = file.createVariable("bmns", np.float64, ("radius", "mn_mode_nyq"))
+        bmns.long_name = "sin(m*t-n*p) component of |B|, on half mesh"
+        bmns.units = "T"
+    m = fit_transform.basis.modes[:, 1]
+    n = fit_transform.basis.modes[:, 2]
+    data = (
+        data_half["|B|"]
+        .reshape(
+            (grid_half.num_theta, grid_half.num_rho, grid_half.num_zeta), order="F"
+        )
+        .transpose((1, 0, 2))
+        .reshape((grid_half.num_rho, -1), order="F")
+    )
+    x_mn = np.zeros((surfs - 1, m.size))
+    for i in range(surfs - 1):
+        x_mn[i, :] = fit(data[i, :])
+    xm, xn, s, c = ptolemy_identity_rev(m, n, x_mn)
+    bmnc[0, :] = 0
+    bmnc[1:, :] = c
+    if not eq.sym:
+        bmns[0, :] = 0
+        bmns[1:, :] = s
+
+    file.close()
+
+
+def _write_terps_input(  # noqa: C901
+    path,
+    mode_family,
+    surfs,
+    lssl,
+    lssd,
+    M_max,
+    N_min,
+    N_max,
+    M_booz_max,
+    N_booz_max,
+    awall,
+    deltaJp,
+    modelk,
+    nfp,
+    nev,
+    al0,
+):
+    """Write TERPSICHORE input file."""
+    ivac = surfs // 4
+    if (N_max > 8) or (M_max > 16):
+        nj = 150
+        nk = 150
+    elif (N_max > 4) or (M_max > 8):
+        nj = 100
+        nk = 100
+    else:
+        nj = 50
+        nk = 50
+
+    f = open(path, "w")
+
+    f.write("               TERPSICHORE\n")
+    f.write("C\n")
+    f.write("C        MM  NMIN  NMAX   MMS NSMIN NSMAX NPROCS INSOL\n")
+    f.write(
+        "         {:>2d}   {:>3d}   {:>3d}    55   {:>3d}   {:>3d}    1     0\n".format(
+            M_booz_max, -N_booz_max, N_booz_max, N_min, N_max
+        )
+    )
+    f.write("C\n")
+    f.write("C        NJ    NK  IVAC  LSSL  LSSD MMAXDF NMAXDF\n")
+    f.write(
+        "        {:>3d}   {:>3d}   {:>3d}  {:>4d}  {:>4d}    120     64\n".format(
+            nj, nk, ivac, lssl, lssd
+        )
+    )
+    f.write("C\n")
+    f.write("C     TABLE OF FOURIER COEFFIENTS FOR BOOZER COORDINATES\n")
+    f.write("C     EQUILIBRIUM SETTINGS ARE COMPUTED FROM FIT/VMEC\n")
+    f.write("C\n")
+
+    boz_str_title = "C M=  0"
+    boz_str_neg = "      0"
+    boz_str_pos = "      1"
+    for m in range(1, 37):
+        if m >= 10:
+            boz_str_title += " " + str(m)[1]
+        else:
+            boz_str_title += " " + str(m)
+        boz_str_neg += " 1"
+        boz_str_pos += " 1"
+    boz_str_title += "  N\n"
+    f.write(boz_str_title)
+
+    for n in range(-N_booz_max, N_booz_max + 1):
+        if n < 0:
+            f.write(boz_str_neg + "{:>3}\n".format(n))
+        else:
+            f.write(boz_str_pos + "{:>3}\n".format(n))
+
+    f.write("C\n")
+    f.write("      LLAMPR      LVMTPR      LMETPR      LFOUPR\n")
```

Comment:
if you konw what these are can you add comments to our code here explaining? or a link to any docs that TERPSICHORE may have to find these definitions?

Reply:
I've added a reference to this documentation page to the `TERPSICHORE` objective function: 
https://princetonuniversity.github.io/STELLOPT/TERPSICHORE.html

Most of these logical flags are either unused or simply control what gets printed during the calculation, so they don't have an actual effect on the result. Hence I hard-coded them.